subprojects/core-api/src/main/java/org/gradle/api/artifacts/repositories/ArtifactResolutionDetails.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.artifacts.repositories;
+
+import org.gradle.api.Action;
+import org.gradle.api.Incubating;
+import org.gradle.api.artifacts.ModuleIdentifier;
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
+import org.gradle.api.attributes.AttributeContainer;
+
+/**
+ * Details about an artifact resolution query. This is used whenever repository
+ * {@link ArtifactRepository#contentFilter(Action)} content filtering is in place.
+ * <p></p>
+ * This interface gives access to the details of the artifact query. There are two
+ * cases when filtering can be called:
+ * <p></p>
+ * <ul>
+ *     <li>when looking for a specific module version, for example org:foo:1.0</li>
+ *     <li>when looking for the list of versions for a module, for example org:foo</li>
+ * </ul>
+ * <p></p>
+ * Listing is called when using dynamic versions (ranges, 1.+, ...).
+ * <p></p>
+ * The module identifier will always be non-null. If what you want to express
+ * is that a module cannot be found in a repository, independently of its version,
+ * then it's enough to just look at the module id using {@link #getModuleId()}.
+ * <p></p>
+ * However, if you have to differentiate depending on the version number (for example,
+ * some versions of a module are found in one repository, others in a different repository),
+ * then you must look at the version thanks to the {@link #getComponentId()} method. But
+ * because there can also be version listings, you must also check for {@link #getModuleId()}.
+ * <p></p>
+ * A {@link #isVersionListing() convenience method} will make it easier to find out if you
+ * are in the version listing case, or module version case.
```

Comment:
I didn't explain this properly but doing this would result in inconsistencies in resolution. It's actually not how the engine works: independently of how the _dependency_ is expressed, the module metadata we get in a single resolution is _unique_. It doesn't depend on the dependency, and in particular it's not because in one place in the tree you find `1.1` and in another you find `[1.0,1.1]` (which both resolve to `1.1`) that they both have different metadata. It's the same, eventually. That's also why we cannot say something like:

```
dependencies {
    implementation("foo:bar:1.0") at "some repo"
}
```
because while it seems to make sense for a single dependency (first level), what does it mean if you transitively find another dependency with the same version? What if you have conflict resolution kicking in? Does it mean that you should try to find the transitive dependencies of "foo:bar:1.0" in the same repository?

Another example is the "status", which I thought could be solved via attributes but cannot: one might thing that setting the "status" attribute to "integration" might be enough not to look for releases for example. However, the status is often only known _once you have parsed metadata_, so you already need to hit the repository before. So, the "consumer attributes" are not the merged attributes of the dependency and the configuration, but can only consider the attributes of the configuration here.

Even answering simple questions like "is it cheap to get metadata for foo:bar:1.0" becomes significantly more complex to implement.

Reply:
I don't get the status argument. If a request says "I want integration versions" and the user knows this is a pure release repo, they can return false and we never have to hit that repo for that request. They don't need any metadata for that, they know their repo never has anything with that status.

Reply:
I think the only thing this API should answer is:
> Given this particular dependency request, is it _possible_ that repository X contains the module.

When we request a dependency from a repository, the request contains:
- A `ModuleComponentSelector`
- A set of consumer attributes.
- (Maybe) a configuration name (although I would much prefer we make it easier to add an attribute than exposing this in the API).

Trying to make the API act on other things that are not part of this request (like the `ModuleComponentIdentifier`) means that we are exposing the way we implement dependency resolution in the API. I really don't think we should do that.